### PR TITLE
Fix issue 15257 - malformed iasm terminates the compiler

### PR DESCRIPTION
**Pull request by Cauterite**
[Issue 15235](https://issues.dlang.org/show_bug.cgi?id=15235) is also fixed incidentally by this patch.

Previously, inline assembly errors would call `exit()` instead of propagating the error to callers. Apparently stack unwinding is not allowed in DMD source, so errors are now propagated manually.

I wasn't exactly sure what to do about calls to `error()`, so for now I've left them as-is. However all calls to `asmerr()` have been augmented.

Mistakes will be made. I am a total git noob, for starters.

### CODE REVIEW

**Comment by Cauterite** [comment]
`"iasm.c:761:56: error: 'SCHAR_MAX' was not declared in this scope"`

I don't understand, I didn't even change this part of the code.

**Comment by Cauterite** [comment]
`src/core/thread.d(2418): Error: variable sp used before set`
Now this I can't explain: somehow I've changed the compiler in such a way that using `= void` variables before writing to them is now an error. See [here](https://github.com/Cauterite/druntime/commit/d57631842905b3c9bedb549d465b8afcb673d5d5) the lines I had to change to make it compile. If anyone could possibly speculate about why this has happened that'd be great, because I simply don't know enough about the compiler to have any clue.

File: src/iasm.c [hunk 1]
```diff
@@ -426,14 +426,10 @@ static OPND *asm_and_exp();
 static OPND *asm_cond_exp();
 static opflag_t asm_determine_operand_flags(OPND *popnd);
 static code *asm_genloc(Loc loc, code *c);
-static int asm_getnum();
+static long long asm_getnum();
```
Comment (UplinkCoder) [inline src/iasm.c]: Why not leave it as int ?

File: src/iasm.c [hunk 2]
```diff
@@ -481,19 +478,23 @@ static void asm_chktok(TOK toknum, const char *msg)
...
+        return false;
     }
+    return true;
 }
 
 
 /*******************************
  */
 
+/* returns PTRNTAB(NULL) on failure */
```
Comment (UplinkCoder) [inline src/iasm.c]: make it a ddoc comment /* > /**

File: src/iasm.c [hunk 3]
```diff
@@ -2342,40 +2391,50 @@ static void asm_merge_symbol(OPND *o1, Dsymbol *s)
             if (ei)
             {
                 o1->disp = ei->exp->toInteger();
-                return;
+                return true;
             }
         }
         if (v->isThreadlocal())
+        {
```
Comment (UplinkCoder) [inline src/iasm.c]: uneeded braces

File: src/iasm.c (hunk 1 above)
Comment (Cauterite) [inline src/iasm.c]: because it needs to be able to return a value outside the `int` range to indicate failure.

Comment (UplinkCoder) [inline src/iasm.c]: -1 would not do ?

Comment (Cauterite) [inline src/iasm.c]: I dunno, you tell me whether `ctfeInterpret()->toInteger()` can return `int(-1)` as a valid result.

**Comment by Cauterite** [comment]
**Status update:** This patch is so close to working. DMD, DRuntime and Phobos are all buildable.

This is the current failure on Linux/BSD:
> posix.mak:475: recipe for target 'checkwhitespace' failed
>make: *** [checkwhitespace] Error 1

I would love it if checkwhitespace would actually tell me why it's failing. I don't have Linux or BSD so I can't run it myself.

On Win32 this is where we're at:
> Makefile:205: recipe for target 'test_results/d_do_test.exe' failed
> gmake: *** [test_results/d_do_test.exe] Error 127

`d_do_test.exe` is crashing on a null dereference. I believe the cause is [this line](https://github.com/dlang/druntime/blob/master/src/rt/cast_.d#L68), where the `typeid(o)` is trying to read a null from the vtable or something. I would expect this is some kind of codegen issue, but I have no clue how my patch has caused it. The error is so far removed from the cause that I don't even know where to start in debugging this.

File: src/iasm.c (hunk 1 above)
Comment (andralex) [inline src/iasm.c]: Hmmm this is not portable to systems where `sizeof(long long) == sizeof(int)`. Should we worry?

Comment (Cauterite) [inline src/iasm.c]: I have a better idea. Since `asm_getnum` is only used in one place, and its output is immediately checked by `isPowerOf2`, we can just return `int(-1)` and the caller will reject it in any case (because `0xffffffff` is not a power of 2).

**Comment by UplinkCoder** [comment]
@Cauterite The observed issues occur when data-layout changes and header-files get out of sync.

**Comment by Cauterite** [comment]
@UplinkCoder thanks, I'll check now for any discrepancies.

**Comment by UplinkCoder** [comment]
@Cauterite try not to change any types.

**Comment by Cauterite** [comment]
As far as I can see, the only types I've changed are the returns of the functions in `iasm.c`, which are static functions. I certainly haven't changed any structure/class layouts.

File: src/iasm.c (hunk 1 above)
Comment (MartinNowak) [inline src/iasm.c]: No implicit assumptions and strong couplings between different code locations please. At best don't even start to reason like that.
toInteger can return a type that can hold any integer of the target machine, so just use the return type of toInteger d_integer_t, see https://github.com/dlang/dmd/blob/2a1e3592d6d91c0d3eac6725584010ca3061422d/src/expression.h#L225.

File: src/iasm.c [hunk 4]
```diff
@@ -413,6 +413,8 @@ struct OPND
     }
 };
 
+static OPND* const emptyOPND = (OPND*)(-1);
```
Comment (MartinNowak) [inline src/iasm.c]: Null seems like a good choice for a missing operand, what's the distinction to empty? Needs an explanatory comment.

Comment (MartinNowak) [inline src/iasm.c]: Ah you're using null now for failed parses. Please switch the logic, errors are rare, missing operands are normal, but testing for null is slightly cheaper than comparing with -1.
So just introduce errorOPND = -1, which also fits how we do/did that in a couple of other places of the compiler. That also leaves room for more error codes.

File: src/iasm.c [hunk 5]
```diff
@@ -505,34 +508,42 @@ static PTRNTAB asm_classify(OP *pop, OPND *popnd1, OPND *popnd2,
...
-    if (!popnd1)
+    if (popnd1 == emptyOPND)
     {
         usNumops = 0;
     }
     else
     {
         popnd1->usFlags = opflags1 = asm_determine_operand_flags(popnd1);
-        if (!popnd2)
+        if (!opflags1)
```
Comment (MartinNowak) [inline src/iasm.c]: Why can't the opflags be 0? That's somewhat uncommon for flags and warrants more explicit code, e.g. if (opflags == OPFLAGSerror) or so.

**Review by MartinNowak — CHANGES_REQUESTED** [review]
Just deal with the segfault as with any reproducible segfault, build a debug dmd, run the program in a debugger, see where it stops, and fix whatever you broke.

**Comment by MartinNowak** [comment]
Ping, this would be a useful addition and the PR is fairly ready. Would be great to finish the last few issues.

File: src/statementsem.d [hunk 6]
```diff
@@ -3363,14 +3363,25 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
...
 
     override void visit(CompoundAsmStatement cas)
     {
         foreach (ref s; *cas.statements)
         {
-            s = s ? s.semantic(sc) : null;
+            if (s)
+            {
+                s = s.semantic(sc);
+                if (cast(AsmStatement) s &&
```
Comment (UplinkCoder) [inline src/statementsem.d]: This will not do a type-check

**Comment by Cauterite** [comment]
closing until i have time to work on this
